Remove celery.task references in modules, docs

ferdnyc opened the pull request:
*Note*: Before submitting this pull request, please review our [contributing
guidelines](https://docs.celeryproject.org/en/master/contributing.html).

## Description

This PR attempts to clean up lingering references to `celery.task` in both the Celery module code and accompanying documentation (excepting previous release notes and etc.), primarily to prevent failures when attempting to display in-module documentation using `pydoc`.

There are a few references in the documentation to modules previously found below `celery.task` (`celery.task.ping`, for example), for which I couldn't find adequate replacements. I've tagged them in this PR by adding `XXX:` comments, to direct attention to those sections of (example) code.

Fixes #6859

ferdnyc commented on docs/userguide/application.rst:
```diff
@@ -475,7 +468,7 @@ chain breaks:
     .. code-block:: python
 
-        from celery.task import task
+        from celery import task
```
Whoops, that's wrong. Will fix.

ferdnyc commented on docs/userguide/periodic-tasks.rst:
```diff
@@ -106,19 +106,19 @@ beat schedule list.
 Setting these up from within the :data:`~@on_after_configure` handler means
-that we'll not evaluate the app at module level when using ``test.s()``. Note that 
+that we'll not evaluate the app at module level when using ``test.s()``. Note that
 :data:`~@on_after_configure` is sent after the app is set up, so tasks outside the
-module where the app is declared (e.g. in a `tasks.py` file located by 
-:meth:`celery.Celery.autodiscover_tasks`) must use a later signal, such as 
+module where the app is declared (e.g. in a `tasks.py` file located by
+:meth:`celery.Celery.autodiscover_tasks`) must use a later signal, such as
```
Sorry, my editor automatically removes trailing whitespace on save.

thedrow commented on docs/internals/guide.rst:
```diff
@@ -186,6 +186,7 @@ Here's an example using Celery in single-mode:
 
 .. code-block:: python
 
+    # XXX: No longer valid?
```
Yes this is no longer valid.

ferdnyc replied:
That's what I figured. Is the "here's the same thing..." version still valid? Should I just rip out the first version and make that second sample the one-and-only?

ferdnyc replied:
(I'll proceed on the assumption that both answers are "yes", and add a commit as described. It's easy enough to drop or tweak further if need be.)

ferdnyc replied:
Hrm. Actually, reading the intro to this section, the whole point is to illustrate the difference between (as the heading _says_, point-blank): "Applications vs. 'single mode'".So it needs two examples, if illustrating that dichotomy is still desirable at all. If not, I can just pull the entire section out.

If it _is_ worth illustrating, it'll need a new first (single-mode) example. Can the current Tasks API still be used single-mode style; is it just a matter of updating the imports?

ferdnyc replied:
@thedrow / @auvipy Any thoughts on this? I believe it's the last outstanding issue. For context, this is the documentation section that reads:

> #### Applications vs. “single mode”
> In the beginning Celery was developed for Django, simply because this enabled us get the project started quickly, while also having a large potential user base.
> 
> In Django there’s a global settings object, so multiple Django projects can’t co-exist in the same process space, this later posed a problem for using Celery with frameworks that don’t have this limitation.
> 
> Therefore the app concept was introduced. When using apps you use ‘celery’ objects instead of importing things from Celery sub-modules, this (unfortunately) also means that Celery essentially has two API’s.
> 
> Here’s an example using Celery in single-mode:
> 
> ```
> from celery import task
> from celery.task.control import inspect
> 
> from .models import CeleryStats
> 
> @task
> def write_stats_to_db():
>     stats = inspect().stats(timeout=1)
>     for node_name, reply in stats:
>         CeleryStats.objects.update_stat(node_name, stats)
> 
> ```
> and here’s the same using Celery app objects:
> 
> ```
> from .celery import celery
> from .models import CeleryStats
> 
> @app.task
> def write_stats_to_db():
>     stats = celery.control.inspect().stats(timeout=1)
>     for node_name, reply in stats:
>         CeleryStats.objects.update_stat(node_name, stats)
> ```
> 
> In the example above the actual application instance is imported from a module in the project, this module could look something like this:
> 
> ```
> from celery import Celery
> 
> app = Celery(broker='amqp://')
> 
> ```

The issue being that the first example is no longer valid code, since the removal of `celery.task`. I am not qualified to write a replacement snippet (or even to determine whether it's still necessary to have one), but I'll be happy to paste one in to the docs if someone can supply it.

ferdnyc replied:
I removed the added comment I'd placed here, as well. Either the first example will get updated, or it won't; since it's only a piece of example code in the documentation, it doesn't affect anything _operationally_ regardless, so this conversation can serve as indication that it needs to be addressed. (Unless someone feels like migrating it to a new Issue.)

thedrow commented on docs/internals/protocol.rst:
```diff
@@ -292,6 +292,8 @@ format:
 
 .. code-block:: javascript
 
+    // XXX: No longer applicable?
```
The response is still the same but the ping task is elsewhere.

ferdnyc replied:
Not sure I understand... is the response _exactly_ the same, or is it a same-except-the-`"task"`-value-changes sort of sameness? If the former I'll just delete my comment, if the latter... perhaps someone can supply an updated JSON response snippet?

thedrow commented on docs/userguide/application.rst:
```diff
@@ -475,16 +468,16 @@ chain breaks:
 ==============
 
-All tasks created using the :meth:`~@task` decorator
+All tasks created using the :meth:`~@app.task` decorator
```
Does this render properly?

f0xxx1 replied:
It does render properly, just tested it.

ferdnyc replied:
Yeah, I stole that from elsewhere in the docs. :grin:

thedrow reviewed (changes requested):
This looks great overall. Thank you for helping us making the documentation more accurate. :heart:

thedrow commented:
Could you please address the rest of the review's comments so we can merge this?
Thanks!

thedrow commented:
@ferdnyc Ping?

ferdnyc commented:
@thedrow Whoops! Thanks, missed the prev notifications somehow. I'll take a look now, sorry for the delay.

auvipy reviewed (changes requested):
can you please check the readthedocs failures?

ferdnyc commented:
@auvipy Yeah, I started looking at those yesterday, and ended up down a rabbit hole of Sphinx intricacies. I hate their documentation, and never seem to get the information I actually need out of it.

I'm not at all sure, now, that the decorator references in the documentation are being handled correctly. Not merely the ones I changed, but all of them. In fact, I have concerns about a lot of the cross-references in the docs, and feel like I may have opened a can of worms by touching them. If I'm right, the issues go deeper than just my changes.

It seems that in the celery docs, references can sometimes be preceded by an `@` character, which has some special meaning. Most of these work, even though I can find nothing in the Sphinx docs that supports such a usage.

However, sometimes it breaks. See, for example, the `add_to_chord` docstring:
https://github.com/celery/celery/blob/ebeb4a4607d83cb5668fad5aaac5d5d8f2fb05b4/celery/app/task.py#L936-L947

That type [actually comes out as a **mailto:** link in the rendered documentation](https://docs.celeryproject.org/en/master/reference/celery.app.task.html#celery.app.task.Task.add_to_chord) because of the `@` sign, instead of linking to `celery.Signature` as presumably intended. Changing it to `~celery.Signature` causes it to display as `Signature` and link to the `celery.Signature` class docs in `reference/celery.html`, which seems to be the desired outcome. So I'm completely confused by all these `@` signs.

Meanwhile, coming back to the decorators, Sphinx actually does support both [`py:decorator`](https://www.sphinx-doc.org/en/master/usage/restructuredtext/domains.html#directive-py-decorator) and [`py:decoratormethod`](https://www.sphinx-doc.org/en/master/usage/restructuredtext/domains.html#directive-py-decoratormethod) domains, for writing decorator documentation. I... can't tell, honestly, if those are actually being used for the decorator documentation, because that's generated by autodoc. Autodoc _does_ (as of version 2.0) have an [`autodecorator`](https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html#directive-autodecorator) directive which I assume makes the answer to that question "yes", but also apparently comes with some caveats:
> #### Note
>
> If you document decorated functions or methods, keep in mind that autodoc retrieves its docstrings by importing the module and inspecting the `__doc__` attribute of the given function or method. That means that if a decorator replaces the decorated function with another, it must copy the original `__doc__` to the new function.

And, as a commenter on [this stackoverflow answer specifically related to celery](https://stackoverflow.com/a/15693082/200794) mentions, documenting the way it's suggested in that answer can cause warnings about duplicate descriptions. The celery docs currently abound with such warnings. A build of all docs with a clean output directory results in 26 instances of 'WARNING: duplicate object description' being emitted.

Even if decorators are documented as decorators, they're still referenced with the `:func:` or `:meth:` inline roles. The Sphinx docs even [take the time to explicitly point that out](https://www.sphinx-doc.org/en/master/usage/restructuredtext/domains.html#directive-py-decorator) (see the bottom of the entry). But they never address the elephant in the room, which is that a <code>:meth:&grave;some.py.function&grave;</code> reference will be formatted as `some.py.function()` and a <code>:meth:&grave;~other.Class.method&grave;</code> reference will format as `method()`. It would be _extremely useful_ to have a <code>:decor:&grave;~some.decorator&grave;</code> role, which would then format in the output as `@decorator`. But maybe that's automatic when the decorators are explicitly documented as decorators. I have no idea, because as  mentioned the Sphinx docs are turrible.

Putting _allllll_ that aside, the crashes in building the docs reference a function, `maybe_resolve_abbreviations`, that I can't seem to find anywhere:
```
Extension error (celerydocs):
Handler <function maybe_resolve_abbreviations at 0x7fe13797a8b0> for event 'missing-reference' threw an exception (exception: 'Unknown abbreviation: @app.task (meth)')
make: *** [Makefile:55: html] Error 2
```

So, the intuition that <code>:meth:&grave;~@app.task&grave;</code> might not format right was spot-on. If  replace all of those references (including the preexisting ones I didn't touch) with something else, the crashes go away and the docs build. The actual effects, however, vary:
* If I use <code>:meth:&grave;app.task&grave;`, the output is **app.task()** and doesn't link anywhere.
* If I use <code>:meth:&grave;~celery.app.task&grave;` or <code>:func:&grave;~celery.app.task&grave;` the output is **task()** and it links to the `celery.app.task` module documentation, which is presumably not the correct location.
* If I use <code>:func:&@task&grave;</code>, the output is **app.task()** (for reasons I cannot comprehend), and it links to the decorator documentation in `celery.Celery.task` (also for reasons I cannot comprehend).

So, presumably the best course of action would be to use <code>:func:&@task&grave;</code> when referencing the `@app.task` decorator, and <code>:func:&~@task&grave;</code> when referencing the `@task` decorator, which will name them both accordingly (except for the `()`-instead-of-`@` formatting issue). They'll both point to the same documentation, which... I _think_... is... right? Honestly at this point I'm completely confused and feeling like I didn't sign up for this.

ferdnyc commented:
I've updated the changes to use <code>:meth:&grave;@task&grave:</code> anywhere **app.task()** should be displayed as a link to the decorator documentation, even though that feels way too magical for my taste. I've also added a commit that fixes those mailto: links from `~@Signature`. I expect the docs will build correctly now; we shall see.
